### DISCUSSION TITLE
How to compile libvips as static binary

### DISCUSSION THREAD

**Discussion by hanssonrickard** [libvips/libvips#3238] [Q&A]
Hi.
I am trying to compile libvips as static (--default-library static), but after it has been built and i do a "file vips", it tells me that it is dynamically linked. I can also verify that by zipping up my vips folder and move it over to another ubuntu machine and try to execute the vips binary and i get an error saying 
"./vips: error while loading shared libraries: libglib-2.0.so.0: cannot open shared object file: No such file or directory"

Or am i just thinking wrong here? I thought the "static" option would make it pretty much movable from my docker where i built it over to another ubuntu machine running same version of OS?

**Comment by jcupitt** [comment]
Hi again,

That'll build libvips itself as a static library, but unless you have static libraries available for all dependencies (glib, libjpeg, etc.) it'll still be dynamically linked to those. And even then you'll probably still want it to be dynamically linked to at least libc, or threading will break and you'll be unable to link with other libraries.

In my opinion static libraries on linux are mostly useless, and have been useless for more than 10 years. Perhaps there's some other solution to your problem? What are you trying to achieve?

**Comment by hanssonrickard** [comment]
Ok, thanks for the answer.
It was just a misunderstanding in that case from my side. I thought it was possible to pretty much make it "portable" so that if i used a machine here to compile it, then it would be easy to move over to a customer machine without having to install a lot of dependencies to make it run.

A follow up question then:
When compiling we install a lof of "dev" packages, but when it is compiled and we wish to run it on another machine, should we install the dev packages on that machine as we ll (as. i know they will install all dependencies needed), or how do I easily know which are the actual dependencies needed to run it on the second machine?

**Reply by jcupitt** [reply]
For binaries, you don't need the dev packages, just the libs.

Python is a bit different: it needs the dev packages when you run `setup.py`, since that's the point at which it generates and builds the shim.

I would do it in docker:

1. Start with a standard install of ubuntu 22.04 or whatever.
2. Install the build tools with `RUN apt install xxx` etc.
3. Install the dev versions of libraries which are already in your base install, so probably libjpeg-dev, libglib-dev, libpng-dev, etc. 
4. Build all non-standard dependencies from source, and install to a private prefix, perhaps `/usr/local/vips`. Hopefully there won't be too many.
5. Finally build libvips to that prefix as well.
6. Clean the install area (you can probably remove the headers, for example, if you're not using python) and tar it up.
7. You now have a set of libs which work together and which only depend on libs in the base OS. This ought to work on any ubuntu 22.04 machine.

This is how the binaries for lambda etc are made.

**Comment by hanssonrickard** [comment]
Ok, thanks a lot for the answer.

Basically, if I only use the standard libs from the ubuntu reps, I should even be able to wrap it up into my own deb package with dependencies defined.


---
Med vänliga hälsningar | Best Regards

Rickard Hansson
System Consultant, NI Solutions
+46 709 34 77 96
www.nisolutions.se<http://www.nisolutions.se>


[image001.jpg]

23 dec. 2022 kl. 14:58 skrev John Cupitt ***@***.***>:



For binaries, you don't need the dev packages, just the libs.

Python is a bit different: it needs the dev packages when you run setup.py, since that's the point at which it generates and builds the shim.

I would do it in docker:

  1.  Start with a standard install of ubuntu 22.04 or whatever.
  2.  Install the build tools with RUN apt install xxx etc.
  3.  Install the dev versions of libraries which are already in your base install, so probably libjpeg-dev, libglib-dev, libpng-dev, etc.
  4.  Build all non-standard dependencies from source, and install to a private prefix, perhaps /usr/local/vips. Hopefully there won't be too many.
  5.  Finally build libvips to that prefix as well.
  6.  Clean the install area (you can probably remove the headers, for example, if you're not using python) and tar it up.
  7.  You now have a set of libs which work together and which only depend on libs in the base OS. This ought to work on any ubuntu 22.04 machine.

This is how the binaries for lambda etc are made.
